Lecture: AI I - Basics 

Previous:
[**Chapter 1.1: Jupyter Notebook Basics**](../01_prerequisites/01_basics.ipynb)

---

# Chapter 2.1: Python Basics

- [Python](#Python)
- [Variables](#Variables)
- [Basic data types](#Basic-data-types)
    - [Boolean type and operators](#Boolean-type-and-operators)
    - [Numeric types and operators](#Numeric-types-and-operators)
    - [String type and operators](#String-type-and-operators)
- [String formatting](#String-formatting)
- [Built-in functions](#Built-in-functions)

## Python
### The Zem of Python

The Zen of Python, authored by Tim Peters, is a collection of guiding principles that influence Python’s design philosophy and programming style. It emphasises simplicity, readability, and clarity. These principles encourage developers to write clean, understandable, and maintainable code rather than overly clever or complicated solutions. You can view the full Zen by running import this in a Python interpreter, reminding yourself of these best practices whenever writing Python code for data analysis, AI, or general software development.

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


### Python version

Python is a versatile, high-level programming language widely used in data analysis, AI, and software development. It has evolved significantly since its first official release, Python 1.0 in 1994, which established its core features. In 2000, Python 2.0 was released. The last version of the Python 2.x series, was Python 2.7 with a end of life support until January 1, 2020. Python 3.0, released in December 2008, which brought backward-incompatible changes to improve consistency, such as treating print as a function and supporting Unicode by default. The current version of Python is 3.13 released in 2024.

For this module, ensure your environment uses Python 3.12 to remain compatible with all examples and exercises.

In [2]:
!python --version

Python 3.12.11


## Variables

In Python, variables are names that store data values. You can assign any value to a variable using the equals sign (`=`), and the type is determined automatically based on what you assign (this is called _dynamic typing_):

In [4]:
a = 1
b = 3.141
c = "Hello, World!"
d = True

a, b, c, d

(1, 3.141, 'Hello, World!', True)

You can assign multiple variables at the same time, for example:

In [6]:
a = b = 10
# Or
a, b = 10, 20

a, b

(10, 20)

## Basic data types

### Numeric type and operators

Python has three main [numeric types](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex): 
- integers (int), which represent whole numbers like 5 or -3
- floating-point numbers (float), which represent real numbers with decimals like 3.14 or -0.01
- complex numbers (complex), used for calculations involving imaginary numbers, written in the form a + bj, such as 2 + 3j 

These types allow you to perform a wide range of mathematical operations easily in Python.

In [7]:
type(1)

int

In [8]:
type(1.1)

float

In [9]:
type(1j)

complex

#### Operation

#### Notation for binary, octal, and hexadecimal numbers

In Python, you can write numbers in different bases using specific prefixes: 
- binary with 0b (e.g. 0b1010 for 10)
- octal with 0o (e.g. 0o12 for 10)
- hexadecimal with 0x (e.g. 0xA for 10)

These notations are useful when working with low-level programming, bitwise operations, or interpreting data in various numeric systems.

In [20]:
0b10101010, int("0b10101010", 2), bin(170)

(170, 170, '0b10101010')

In [21]:
0o252, int("0o252", 8), oct(170)

(170, 170, '0o252')

In [22]:
0xAA, int("0xAA", 16), hex(170)

(170, 170, '0xaa')

# Decimal as additional numeric type

The decimal module in Python provides the [Decimal](https://docs.python.org/3/library/decimal.html) data type for decimal floating-point arithmetic with high precision, which is especially useful for financial calculations where accuracy is critical. Unlike regular float numbers that can have small rounding errors due to binary representation, Decimal ensures precise decimal calculations.

In [24]:
0.1 + 0.2

0.30000000000000004

In [26]:
from decimal import Decimal

Decimal("0.1") + Decimal("0.2")

Decimal('0.3')

### Boolean types and operators

Boolean values in Python represent truth values and can be either `True` or `False`. They are often used in conditional statements and logical operations.

In [19]:
type(True), type(False)

(bool, bool)

Python provides several [operators](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not) for working with boolean values:

| Operation | Meaning |
|----|---|
| `and`  | Logical AND (conjunction)|
| `or`  | Logical OR (disjunction) |
| `not`  | 	Logical NOT (negation) |

In [28]:
True and True, True or False, not False

(True, True, True)

in addition to these logical operators, Python also supports [comparison operators](https://docs.python.org/3/library/stdtypes.html#comparisons) that return boolean values based on the comparison of two values:

| Operation | Meaning |
|----|---|
| `<`  | strictly less than |
| `<=`  | less than or equal |
| `>`  | strictly greater than |
| `>=`  | greater than or equal |
| `==`  | equal |
| `!=`  | not equal |
| `is`  | object identity |
| `is not`  | negated object identity |
| `in`  | membership test (checks if a value is in a collection) |
| `not in`  | negated membership test (checks if a value is not in a collection) |



In [33]:
print(1 < 2)
print(1 <= 1)
print(1 > 0)
print(1 >= 1)
print(1 == 1)
print(1 != 2)

True
True
True
True
True
True


In [37]:
print(3 > 2 > 1)
print(1 < 2 < 3)

True
True


### String type and operators

In [11]:
type("Hello, World!")

str

## String formatting

## Built-in functions

---

Lecture: AI I - Basics 

Next: [**Chapter 2.2: Data Structures**](../02_python/02_data_structures.ipynb)